<a href="https://colab.research.google.com/github/novaladriann/pm-turi2-preprocessing-noval-adrian/blob/main/PM_P4_Noval_Adrian_2488010018.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

# Dataset karyawan (sengaja mengandung masalah kualitas)
data = {
    'usia'      : [25, 32, np.nan, 45, 28, 51, 38, np.nan, 29, 41],
    'pendapatan': [4_000_000, 7_500_000, 5_200_000, 12_000_000, 4_800_000,
                   15_000_000, 8_100_000, 6_300_000, np.nan, 9_900_000],
    'pendidikan': ['SMA','S1','SMA','S2','SMA','S2','S1','S1','SMA','S2'],  # ordinal
    'kota'      : ['Bandung','Jakarta','Bandung','Surabaya','Jakarta',
                   'Surabaya','Jakarta','Bandung','Bandung','Jakarta'],    # nominal
    'membeli'   : ['Tidak','Ya','Tidak','Ya','Tidak','Ya','Ya','Tidak','Tidak','Ya']  # target
}
df = pd.DataFrame(data)
df

,usia,pendapatan,pendidikan,kota,membeli
0,25.0,4000000.0,SMA,Bandung,Tidak
1,32.0,7500000.0,S1,Jakarta,Ya
2,NaN,5200000.0,SMA,Bandung,Tidak
3,45.0,12000000.0,S2,Surabaya,Ya
4,28.0,4800000.0,SMA,Jakarta,Tidak
5,51.0,15000000.0,S2,Surabaya,Ya
6,38.0,8100000.0,S1,Jakarta,Ya
7,NaN,6300000.0,S1,Bandung,Tidak
8,29.0,NaN,SMA,Bandung,Tidak
9,41.0,9900000.0,S2,Jakarta,Ya


In [ ]:
print("Jumlah nilai hilang per kolom:")
print(df.isna().sum())
print("\nInfo tipe data:")
df.info()

Jumlah nilai hilang per kolom:
usia          2
pendapatan    1
pendidikan    0
kota          0
membeli       0
dtype: int64

Info tipe data:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   usia        8 non-null      float64
 1   pendapatan  9 non-null      float64
 2   pendidikan  10 non-null     object 
 3   kota        10 non-null     object 
 4   membeli     10 non-null     object 
dtypes: float64(2), object(3)
memory usage: 532.0+ bytes


### Langkah 1 — Menangani nilai hilang (imputasi median/modus).

In [ ]:
df['usia'] = df['usia'].fillna(df['usia'].median())
df['pendapatan'] = df['pendapatan'].fillna(df['pendapatan'].median())

print("Jumlah nilai hilang setelah imputasi:")
print(df.isna().sum())
df

Jumlah nilai hilang setelah imputasi:
usia          0
pendapatan    0
pendidikan    0
kota          0
membeli       0
dtype: int64


,usia,pendapatan,pendidikan,kota,membeli
0,25.0,4000000.0,SMA,Bandung,Tidak
1,32.0,7500000.0,S1,Jakarta,Ya
2,35.0,5200000.0,SMA,Bandung,Tidak
3,45.0,12000000.0,S2,Surabaya,Ya
4,28.0,4800000.0,SMA,Jakarta,Tidak
5,51.0,15000000.0,S2,Surabaya,Ya
6,38.0,8100000.0,S1,Jakarta,Ya
7,35.0,6300000.0,S1,Bandung,Tidak
8,29.0,7500000.0,SMA,Bandung,Tidak
9,41.0,9900000.0,S2,Jakarta,Ya


### Langkah 2 — Memisahkan fitur (X) dan label (y).

In [ ]:
X = df.drop(columns=['membeli'])
y = df['membeli']

print("Fitur (X):")
display(X)
print("\nLabel (y):")
print(y)

Fitur (X):


,usia,pendapatan,pendidikan,kota
0,25.0,4000000.0,SMA,Bandung
1,32.0,7500000.0,S1,Jakarta
2,35.0,5200000.0,SMA,Bandung
3,45.0,12000000.0,S2,Surabaya
4,28.0,4800000.0,SMA,Jakarta
5,51.0,15000000.0,S2,Surabaya
6,38.0,8100000.0,S1,Jakarta
7,35.0,6300000.0,S1,Bandung
8,29.0,7500000.0,SMA,Bandung
9,41.0,9900000.0,S2,Jakarta



Label (y):
0    Tidak
1       Ya
2    Tidak
3       Ya
4    Tidak
5       Ya
6       Ya
7    Tidak
8    Tidak
9       Ya
Name: membeli, dtype: object


### Langkah 3 — Encoding kategorikal (ordinal & one-hot).

In [ ]:
X['pendidikan'] = X['pendidikan'].map({'SMA':0,'S1':1,'S2':2})  # ordinal
X = pd.get_dummies(X, columns=['kota'], dtype=int)                 # nominal
y = y.map({'Tidak':0,'Ya':1})

print("X setelah encoding:")
display(X)
print("\ny setelah encoding:")
print(y)

X setelah encoding:


,usia,pendapatan,pendidikan,kota_Bandung,kota_Jakarta,kota_Surabaya
0,25.0,4000000.0,0,1,0,0
1,32.0,7500000.0,1,0,1,0
2,35.0,5200000.0,0,1,0,0
3,45.0,12000000.0,2,0,0,1
4,28.0,4800000.0,0,0,1,0
5,51.0,15000000.0,2,0,0,1
6,38.0,8100000.0,1,0,1,0
7,35.0,6300000.0,1,1,0,0
8,29.0,7500000.0,0,1,0,0
9,41.0,9900000.0,2,0,1,0



y setelah encoding:
0    0
1    1
2    0
3    1
4    0
5    1
6    1
7    0
8    0
9    1
Name: membeli, dtype: int64


### Langkah 4 — Membagi data (SEBELUM scaling).

In [ ]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(
X, y, test_size=0.3, random_state=42)

print("Ukuran X_train:", X_train.shape)
print("Ukuran X_test :", X_test.shape)
display(X_train)

Ukuran X_train: (7, 6)
Ukuran X_test : (3, 6)


,usia,pendapatan,pendidikan,kota_Bandung,kota_Jakarta,kota_Surabaya
0,25.0,4000000.0,0,1,0,0
7,35.0,6300000.0,1,1,0,0
2,35.0,5200000.0,0,1,0,0
9,41.0,9900000.0,2,0,1,0
4,28.0,4800000.0,0,0,1,0
3,45.0,12000000.0,2,0,0,1
6,38.0,8100000.0,1,0,1,0


### Langkah 5 — Penskalaan (fit pada latih, transform pada latih & uji).

In [ ]:
from sklearn.preprocessing import StandardScaler

num = ['usia', 'pendapatan', 'pendidikan']
sc = StandardScaler()
X_train[num] = sc.fit_transform(X_train[num])
X_test[num] = sc.transform(X_test[num])

print("X_train setelah scaling:")
display(X_train)
print("\nX_test setelah scaling:")
display(X_test)

X_train setelah scaling:


,usia,pendapatan,pendidikan,kota_Bandung,kota_Jakarta,kota_Surabaya
0,-1.588665,-1.169416,-1.028992,1,0,0
7,-0.044130,-0.325129,0.171499,1,0,0
2,-0.044130,-0.728919,-1.028992,1,0,0
9,0.882592,0.996363,1.371989,0,1,0
4,-1.125304,-0.875751,-1.028992,0,1,0
3,1.500406,1.767234,1.371989,0,0,1
6,0.419231,0.335617,0.171499,0,1,0



X_test setelah scaling:


,usia,pendapatan,pendidikan,kota_Bandung,kota_Jakarta,kota_Surabaya
8,-0.970851,0.115368,-1.028992,1,0,0
1,-0.507490,0.115368,0.171499,0,1,0
5,2.427127,2.868478,1.371989,0,0,1


### Langkah 6 — Verifikasi: tidak ada nilai hilang, semua numerik, skala seragam.

In [ ]:
print("Nilai hilang di X_train:\n", X_train.isna().sum())
print("\nNilai hilang di X_test:\n", X_test.isna().sum())
print("\nTipe data X_train:\n", X_train.dtypes)
print("\nStatistik kolom numerik X_train (mean seharusnya ~0, std ~1):")
display(X_train[num].describe())

Nilai hilang di X_train:
 usia             0
pendapatan       0
pendidikan       0
kota_Bandung     0
kota_Jakarta     0
kota_Surabaya    0
dtype: int64

Nilai hilang di X_test:
 usia             0
pendapatan       0
pendidikan       0
kota_Bandung     0
kota_Jakarta     0
kota_Surabaya    0
dtype: int64

Tipe data X_train:
 usia             float64
pendapatan       float64
pendidikan       float64
kota_Bandung       int64
kota_Jakarta       int64
kota_Surabaya      int64
dtype: object

Statistik kolom numerik X_train (mean seharusnya ~0, std ~1):


,usia,pendapatan,pendidikan
count,7.000000e+00,7.000000e+00,7.000000e+00
mean,1.110223e-16,1.030921e-16,-1.982541e-17
std,1.080123e+00,1.080123e+00,1.080123e+00
min,-1.588665e+00,-1.169416e+00,-1.028992e+00
25%,-5.847169e-01,-8.023348e-01,-1.028992e+00
50%,-4.412958e-02,-3.251291e-01,1.714986e-01
75%,6.509113e-01,6.659903e-01,7.717436e-01
max,1.500406e+00,1.767234e+00,1.371989e+00


---
## Latihan Mandiri

### 1. Ganti imputasi usia menjadi mean; bandingkan hasilnya.

In [ ]:
df_mentah = pd.DataFrame(data)

median_usia = df_mentah['usia'].median()
mean_usia = df_mentah['usia'].mean()
print(f"Median usia: {median_usia}")
print(f"Mean usia: {mean_usia:.2f}")

perbandingan = pd.DataFrame({
    'usia_asli': df_mentah['usia'],
    'imputasi_median': df_mentah['usia'].fillna(median_usia),
    'imputasi_mean': df_mentah['usia'].fillna(mean_usia)
})
perbandingan

Median usia: 35.0
Mean usia: 36.12


,usia_asli,imputasi_median,imputasi_mean
0,25.0,25.0,25.000
1,32.0,32.0,32.000
2,NaN,35.0,36.125
3,45.0,45.0,45.000
4,28.0,28.0,28.000
5,51.0,51.0,51.000
6,38.0,38.0,38.000
7,NaN,35.0,36.125
8,29.0,29.0,29.000
9,41.0,41.0,41.000


**Perbandingan hasil:** Median usia adalah 35.0, sedangkan mean usia sekitar 36.13. Karena distribusi usia relatif tidak memiliki outlier ekstrem, kedua metode menghasilkan nilai pengganti yang berdekatan.

### 2. Gunakan MinMaxScaler (0–1) sebagai ganti StandardScaler; tampilkan describe().

In [ ]:
from sklearn.preprocessing import MinMaxScaler

X_mm = df.drop(columns = ['membeli']).copy()
X_mm['pendidikan'] = X_mm['pendidikan'].map({'SMA':0,'S1':1,'S2':2})
X_mm = pd.get_dummies(X_mm, columns=['kota'], dtype=int)

X_train_mm,X_test_mm,y_train_mm,y_test_mm = train_test_split(
X_mm, y, test_size=0.3, random_state=42)

mm = MinMaxScaler()
X_train_mm[num] = mm.fit_transform(X_train_mm[num])
X_test_mm[num] = mm.transform(X_test_mm[num])

print("Describe X_train setelah MinMaxScaler:")
display(X_train_mm[num].describe())

Describe X_train setelah MinMaxScaler:


,usia,pendapatan,pendidikan
count,7.000000,7.000000,7.000000
mean,0.514286,0.398214,0.428571
std,0.349660,0.367808,0.449868
min,0.000000,0.000000,0.000000
25%,0.325000,0.125000,0.000000
50%,0.500000,0.287500,0.500000
75%,0.725000,0.625000,0.750000
max,1.000000,1.000000,1.000000


### 3. Tambahkan kolom kategorikal baru dan terapkan one-hot encoding.

In [ ]:
df_new = df.copy()

df_new['status_pernikahan'] = ['Menikah','Lajang','Lajang','Menikah','Lajang','Menikah','Menikah','Lajang','Lajang','Menikah']

X_new = df_new.drop(columns=['membeli'])
X_new['pendidikan'] = X_new['pendidikan'].map({'SMA':0,'S1':1,'S2':2})
X_new = pd.get_dummies(X_new, columns=['kota','status_pernikahan'], dtype=int)

X_new

,usia,pendapatan,pendidikan,kota_Bandung,kota_Jakarta,kota_Surabaya,status_pernikahan_Lajang,status_pernikahan_Menikah
0,25.0,4000000.0,0,1,0,0,0,1
1,32.0,7500000.0,1,0,1,0,1,0
2,35.0,5200000.0,0,1,0,0,1,0
3,45.0,12000000.0,2,0,0,1,0,1
4,28.0,4800000.0,0,0,1,0,1,0
5,51.0,15000000.0,2,0,0,1,0,1
6,38.0,8100000.0,1,0,1,0,0,1
7,35.0,6300000.0,1,1,0,0,1,0
8,29.0,7500000.0,0,1,0,0,1,0
9,41.0,9900000.0,2,0,1,0,0,1


### 4. Jelaskan mengapa fit_transform pada latih tetapi hanya transform pada uji.

`fit_transform` menghitung parameter penskalaan (mean & std untuk `StandardScaler`, atau min & max untuk `MinMaxScaler`) dari data latih, lalu langsung menerapkannya. Data uji hanya di-`transform` (tanpa `fit` ulang) agar parameter yang dipakai tetap konsisten dengan yang dipelajari dari data latih.

Jika data uji ikut di-`fit`, maka informasi dari data uji (yang seharusnya mensimulasikan data baru/tidak diketahui) akan "bocor" ke proses penskalaan, inilah yang disebut **data leakage**. Data leakage membuat evaluasi model menjadi terlalu optimis karena model secara tidak langsung sudah "mengintip" karakteristik data uji sebelum benar-benar dievaluasi.

### 1. Mengapa urutan 'split dulu, baru scaling' penting?

Jika scaling dilakukan sebelum split, statistik (mean, std, min, max) yang dipakai untuk penskalaan akan dihitung dari seluruh data, termasuk data yang seharusnya menjadi data uji. Akibatnya, informasi data uji ikut memengaruhi proses transformasi data latih, sehingga terjadi data leakage. Dengan split terlebih dahulu, data uji benar-benar merepresentasikan data baru yang belum pernah "dilihat" model maupun proses preprocessing, sehingga evaluasi performa model menjadi lebih jujur dan realistis.

### 2. Kapan label encoding dan kapan one-hot encoding?

- **Label/ordinal encoding** digunakan ketika kategori memiliki **urutan atau tingkatan yang bermakna**, misalnya `pendidikan` (SMA < S1 < S2). Urutan angka hasil encoding merepresentasikan urutan tingkatan tersebut.
- **One-hot encoding** digunakan ketika kategori bersifat **nominal**, tanpa urutan, misalnya `kota` (Bandung, Jakarta, Surabaya). Jika kategori nominal dipaksakan memakai label encoding, model bisa salah menafsirkan adanya hubungan "lebih besar/lebih kecil" antar kategori yang sebenarnya tidak ada.

### 3. Apa perbedaan normalisasi dan standardisasi?
- **Normalisasi** (mis. `MinMaxScaler`) mengubah skala fitur ke rentang tertentu, umumnya [0, 1], berdasarkan nilai minimum dan maksimum data. Cocok saat distribusi data tidak diasumsikan normal dan rentang nilai perlu dibatasi.
- **Standardisasi** (mis. `StandardScaler`) mengubah data agar memiliki mean = 0 dan standar deviasi = 1, berdasarkan distribusi data (mean & std), tanpa membatasi rentang nilai secara eksplisit. Lebih cocok digunakan saat data mendekati distribusi normal atau saat algoritma sensitif terhadap distribusi (mis. regresi linear, SVM, PCA).
